mllib/src/main/scala/org/apache/spark/ml/clustering/KMeans.scala

```diff
@@ -330,10 +334,42 @@ class KMeans @Since("1.5.0") (
   @Since("3.0.0")
   def setWeightCol(value: String): this.type = set(weightCol, value)
 
+  /**
+   * Set block size for stacking input data in matrices.
+   * If blockSize == 1, then stacking will be skipped, and each vector is treated individually;
+   * If blockSize &gt; 1, then vectors will be stacked to blocks, and high-level BLAS routines
+   * will be used if possible (for example, GEMV instead of DOT, GEMM instead of GEMV).
+   * Recommended size is between 10 and 1000. An appropriate choice of the block size depends
+   * on the sparsity and dim of input datasets, the underlying BLAS implementation (for example,
+   * f2jBLAS, OpenBLAS, intel MKL) and its configuration (for example, number of threads).
+   * Note that existing BLAS implementations are mainly optimized for dense matrices, if the
+   * input dataset is sparse, stacking may bring no performance gain, the worse is possible
+   * performance regression.
+   * Default is 1.
+   *
+   * @group expertSetParam
+   */
+  @Since("3.1.0")
+  def setBlockSize(value: Int): this.type = set(blockSize, value)
+  setDefault(blockSize -> 1)
+
   @Since("2.0.0")
   override def fit(dataset: Dataset[_]): KMeansModel = instrumented { instr =>
+
     transformSchema(dataset.schema, logging = true)
+    instr.logPipelineStage(this)
+    instr.logDataset(dataset)
+    instr.logParams(this, featuresCol, predictionCol, k, initMode, initSteps, distanceMeasure,
+      maxIter, seed, tol, weightCol, blockSize)
+
+    if ($(blockSize) == 1) {
+      trainOnRows(dataset)
+    } else {
+      trainOnBlocks(dataset)
+    }
+  }
 
+  private def trainOnRows(dataset: Dataset[_]): KMeansModel = instrumented { instr =>
```

Comment:
Some parts in `trainOnRows` and `trainOnBlocks` are the same, and can be moved to `def fit(dataset: Dataset[_])`;
The `instr` should not be missing;

input type of `trainOnRows` and `trainOnBlocks` can be a `RDD[(Vector, Double)]`